doc/api/modules.md

```diff
@@ -912,6 +912,29 @@ by the [module wrapper][]. To access it, require the `Module` module:
 const builtin = require('module').builtinModules;
 ```
 
+### module.containsModuleSyntax(source)
```

Comment:
Not blocking, but might the method name be better as something like `usesModuleSyntax()` or even `usesEsmSyntax()`? Or `usesESM()`? Once we add the method name, we risk being stuck with it forever (even if ESM is experimental status), so it's probably worth an irksome bikeshed to et the name right.

If the irksome bikeshed already happened in some other issue or whatever, by all means, ignore me.

Reply:
Or if we want to parallel `String.includes`, we could use `includesModuleSyntax`, but that feels less readable to me. I went with `contains` because all this method does is check that the syntax is present in the string; you would think that that would mean therefore that the syntax is “used,” but would it really be “used” if the source code throws a parse or syntax error? Is an `export` statement “used” if nothing is importing what it’s exporting?

As for “Module” versus ESM/ES module or variants, technically this is searching for the syntax particular to the Module parse goal (`import`/`export` statements) and we’re using “module” as the user-facing name for ESM things like in `package.json` `"type": "module"` and `--input-type=module` etc.

Reply:
i would recommend `has%TYPE%Syntax`, because bytes & keystrokes. correlates to e.g. 'Map#has`

Reply:
In my experience, “has” usually refers to collections: “the array has element `'foo'`,” “the object has key `'foo'`.” It sounds odd to use it to describe a substring: “the string has substring `'foo'`.” At least to me, “the string contains substring `'foo'`” sounds more natural. (And this method takes a string as its only input.)

Reply:
i can imagine extensions to the API that take things such as an unresolved module import or a URL object, so i'm not sure i agree that the current implementation detail should hold water for naming. 

from a dictionary standpoint, `has`, `contains` and `includes` are all roughly interchangeable:
- the source HAS module syntax
- the source INCLUDES module syntax
- the source CONTAINS module syntax
...in which case i strongly believe the terser syntax wins hands-down.

as far as `has` usually referring to collections: that isn't my experience at all. at the very least i would prefer to say something along the lines of "the string has 'foo' in it", especially when talking to a new programmer who might still be wrapping their head around what strings actually are, let alone how they relate to substrings. but i admit that last part is a bit of a strawman argument.

Reply:
Contains is the more accurate term in my opinion — it intuits a quality that something that is of a predetermined state is made up of the stuff that in part is exactly what you seek.

has being closer with that respect — but JS has a bias for Sets et al, and includes is furthest imho regardless of the bias factor.

Reply:
Referring to the dictionary, of the 25 definitions of “has” the most relevant is [“to possess, own, hold”](https://en.wiktionary.org/wiki/have#English) as in “She has a house and a car.”

Of the 5 definitions of “contains” the first two are both relevant: “to hold inside,” [“to include as a part”](https://en.wiktionary.org/wiki/contain#English) as in “the candy contains peanuts.”

The situation we have here, of module syntax within source code, falls in the latter category. The module syntax is a part of the source code like the peanuts are part of the candy. If you wanted to use “has” to describe the peanuts in the candy, you wouldn’t write “the candy has peanuts,” you would write “the candy has peanuts in it,” because the relationship between the candy and the peanuts is one of enclosure, not of ownership.

Reply:
@GeoffreyBooth I think that the balance here comes from the orientation of the audience, which will likely bias mostly towards JS concepts, like `set.has`. And while it is fair to note that `string.includes` is also JS, I was trying to make the distinction that `string.includes('…')` is not the abstract function of this utility. Instead, we're more like `ModuleSyntax.test(string)` and that is not equivalent in meaning to either `includes` or `has` and thus the preference for `contains`.

Reply:
That said, I think if we ever have a `String.prototype.somethingInsideOfThisIsOneThatMatchesThe(pattern)` it will likely just be the boring ol'`contains` based on this thread 😉

Reply:
If people don’t like `contains` I think the only other logical choice is `includes`, for parity with `String.prototype.includes`. It’s worth noting that the original name for that was `contains`, and only got renamed because of MooTools: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes#String.prototype.contains.

There’s also good info in https://github.com/tc39/Array.prototype.includes/#why-includes-instead-of-has:

> If you survey existing APIs, `has` is used for conceptual “keys,” whereas `includes` is used for conceptual “values.”

Reply:
I think this might be a [good case for contains](http://www.ecma-international.org/ecma-262/#sec-module-semantics-static-semantics-containsduplicatelabels).

Reply:
But if you look closer:

> Let hasDuplicates be ContainsDuplicateLabels of ModuleItemList with argument labelSet.

Ah well, at least they did not manage to throw includes in there too 😊

Reply:
briefly, i'll point out that [the second definition of `hold`](https://en.wiktionary.org/wiki/hold) is `To contain or store`, which lines up with my argument that they're _roughly interchangeable_.

that being said, i agree that tc39 guidance on terminology should take precedence.

Reply:
@GeoffreyBooth thank you for taking the time & effort to detail your arguments :)